packages/commerce-sdk-react/src/hooks/ShopperCustomers/cache.ts

```diff
@@ -17,81 +19,231 @@ import {
         return {
-            // TODO: Rather than invalidate, can we selectively update?
-            ...invalidateCustomer(newParams),
-            update: [{queryKey: getCustomerAddress.queryKey(newParams)}]
+            update: [
+                {
+                    queryKey: getCustomerAddress.queryKey(newParams)
+                },
+                {
+                    queryKey: getCustomer.queryKey(newParams),
+                    updater: createUpdateFunction((customer: Customer) => {
+                        // Push new address onto the end of addresses list.
+                        customer?.addresses?.push(response)
+
+                        return customer
+                    })
+                }
+            ]
         }
     },
     createCustomerPaymentInstrument(customerId, {parameters}, response) {
         const newParams = {...parameters, paymentInstrumentId: response.paymentInstrumentId}
         return {
-            // TODO: Rather than invalidate, can we selectively update?
-            ...invalidateCustomer(newParams),
-            update: [{queryKey: getCustomerPaymentInstrument.queryKey(newParams)}]
+            update: [
+                {
+                    queryKey: getCustomerPaymentInstrument.queryKey(newParams)
+                },
+                {
+                    queryKey: getCustomer.queryKey(newParams),
+                    updater: createUpdateFunction((customer: Customer) => {
+                        // Push new address onto the end of addresses list.
+                        customer?.paymentInstruments?.push(response)
+
+                        return customer
+                    })
+                }
+            ]
         }
     },
     createCustomerProductList(customerId, {parameters}, response) {
         // We always invalidate, because even without an ID we assume that something has changed
-        // TODO: Rather than invalidate, can we selectively update?
-        const invalidate: CacheUpdate['invalidate'] = [
-            {queryKey: getCustomerProductLists.queryKey(parameters)}
-        ]
-        // We can only update cache for this product list if we have the ID
         const listId = response.id
-        if (!listId) return {invalidate}
         return {
-            invalidate,
-            update: [{queryKey: getCustomerProductList.queryKey({...parameters, listId})}]
+            update: [
+                {
+                    queryKey: getCustomerProductLists.queryKey(parameters),
+                    updater: createUpdateFunction((result: CustomerProductListResult) => {
+                        // Push new address onto the end of addresses list.
+                        // Add new list to front of the lists.
+
+                        // Remove the list from the result object
+                        result.data.unshift(response)
+                        result.limit++
+                        result.total++
```

Comment:
So this might have unintended consequences, depending on how pagination is implemented. If the wishlists view shows 5 wishlists per page, the customer has 5, and then adds 1, this will result in 6 wishlists in the response. Depending on the implementation that might be bad, or it might be no big deal. I dunno, but I figured I'd point it out.

Reply:
So the interesting thing is that pagination on the API side is not implemented I don't think. Seems as tho it's a little half baked, you can't request different pages, but the result is returned in a shape that looks like it would be paginated. There is always a total and a limit, and they are always both equal to the absolute number of items in the list, and all the items in the list are returned.